cli_tools/gce_windows_upgrade/upgrader/workflows.go

```diff
@@ -0,0 +1,216 @@
+
+	cleanupWf, err := daisycommon.ParseWorkflow(cleanupWorkflowPath, cleanupVarMap,
+		params.project, params.zone, params.ScratchBucketGcsPath, params.Oauth, params.Timeout,
+		params.Ce, params.GcsLogsDisabled, params.CloudLogsDisabled, params.StdoutLogsDisabled)
+	if err != nil {
+		return nil, err
+	}
+
+	err = daisyutils.RunWorkflowWithCancelSignal(ctx, cleanupWf)
+	return cleanupWf, err
+}
+
+func rollback(ctx context.Context, params *UpgradeParams, installMediaDiskName, newOSDiskName string) (*daisy.Workflow, error) {
+	originalStartupScriptURL := ""
+	if windowsStartupScriptURLBackup != nil {
+		originalStartupScriptURL = *windowsStartupScriptURLBackup
+	}
+
+	w := &daisy.Workflow{
+		Name:           "rollback",
+		DefaultTimeout: "30m",
+	}
+
+	stepStopInstance, err := newStep(w, "stop-instance")
+	if err != nil {
+		return w, err
+	}
+	stepStopInstance.StopInstances = &daisy.StopInstances{
+		Instances: []string{params.InstanceURI},
+	}
+
+	stepDetachNewOSDisk, err := newStep(w, "detach-new-os-disk", stepStopInstance)
+	if err != nil {
+		return w, err
+	}
+	stepDetachNewOSDisk.DetachDisks = &daisy.DetachDisks{
+		{
+			Instance:   params.InstanceURI,
+			DeviceName: fmt.Sprintf("projects/%v/zones/%v/devices/%v", params.project, params.zone, params.osDiskDeviceName),
+		},
+	}
+
+	stepAttachOldOSDisk, err := newStep(w, "attach-old-os-disk", stepDetachNewOSDisk)
+	if err != nil {
+		return w, err
+	}
+	stepAttachOldOSDisk.AttachDisks = &daisy.AttachDisks{
+		{
+			Instance: params.InstanceURI,
+			AttachedDisk: compute.AttachedDisk{
+				Source:     params.osDisk,
+				DeviceName: params.osDiskDeviceName,
+				AutoDelete: params.osDiskAutoDelete,
+				Boot:       true,
+			},
+		},
+	}
+
+	stepDetachInstallMediaDisk, err := newStep(w, "detach-install-media-disk", stepAttachOldOSDisk)
+	if err != nil {
+		return w, err
+	}
+	stepDetachInstallMediaDisk.DetachDisks = &daisy.DetachDisks{
+		{
+			Instance:   params.InstanceURI,
+			DeviceName: fmt.Sprintf("projects/%v/zones/%v/devices/%v", params.project, params.zone, installMediaDiskName),
+		},
+	}
+
+	stepRestoreScript, err := newStep(w, "restore-script", stepDetachInstallMediaDisk)
+	if err != nil {
+		return w, err
+	}
+	stepRestoreScript.UpdateInstancesMetadata = &daisy.UpdateInstancesMetadata{
+		{
+			Instance: params.InstanceURI,
+			Metadata: map[string]string{
+				"windows-startup-script-url": originalStartupScriptURL,
+			},
+		},
+	}
+
+	stepStartInstance, err := newStep(w, "start-instance", stepRestoreScript)
+	if err != nil {
+		return w, err
+	}
+	stepStartInstance.StartInstances = &daisy.StartInstances{
+		Instances: []string{params.InstanceURI},
+	}
+
+	stepDeleteNewOSDisk, err := newStep(w, "delete-new-os-disk", stepStartInstance)
+	if err != nil {
+		return w, err
+	}
+	stepDeleteNewOSDisk.DeleteResources = &daisy.DeleteResources{
+		Disks: []string{
+			fmt.Sprintf("projects/%v/zones/%v/disks/%v", params.project, params.zone, newOSDiskName),
```

Comment:
generating resource URLs can be extracted to a function. It can accept resource type and it can have variants for global and zonal resources. This would reduce the possibility of errors in URIs generated. There are quite a few places where URIs are generated in this file.

Reply:
done for devices and disks.